### ISSUE TITLE
Trouble on pre-training models on Cityscapes

### ISSUE THREAD

**Issue by UltronAI** [JiawangBian/SC-SfMLearner-Release#26]
I calculated the scale factor and scaled my predicted poses like original SfMLearner, and used the evaluation tool of Depth-VO-Feat (Zhan et al.)
```
scale_factor = np.sum(pred[:, :, -1] * gt[:, :, -1]) / np.sum(pred[:, :, -1] ** 2)
pred[:, :, -1] = pred[:, :, -1] * scale_factor
```
But the final result is so poor on KITTI seq09: 
Average translational RMSE (%):  73.9656283885415
Average rotational error (deg/100m):  4.221253537085405

I used the same training hyper-parameters as in train_resnet_256.sh.

What should I do to fix it? Thanks!

**Comment by JiawangBian** [comment]
1. Use "https://github.com/Huangying-Zhan/kitti-odom-eval" for VO evaluation to avoid wrong scale calculation.
2. If results are still bad, please check whether you model is correctly trained. e.g., You can evaluate the depth model in KITTI test images, and see whether it is reasonable.
3. If depth model is bad, you may check the training and validation loss

**Comment by UltronAI** [comment]
@JiawangBian Thanks for your reply! The results got better in the case where I trained models with gt. But I find there is a big gap between w/t gt and w/o gt, keeping the same other parameters. For example, I evaluate the depth model:

without gt:
abs_rel,     sq_rel,        rms,    log_rms,     d1_all,         a1,         a2,         a3
    0.4429,     4.7569,    12.0832,     0.5876,     0.0000,     0.3033,     0.5608,     0.7662

with gt:
abs_rel,     sq_rel,        rms,    log_rms,     d1_all,         a1,         a2,         a3
    0.1571,     1.1660,     5.8016,     0.2358,     0.0000,     0.7859,     0.9281,     0.9724

I see in other issues that you said your methods could work well when w/o gt for validation. Is there anything wrong I did? Or when I want to train models w/o gt, I need to carefully tune the parameters?

**Comment by JiawangBian** [comment]
It seems that you meet a bug when training w/o gt。you may check the training loss and validation loss。Or you can just train it again。

**Comment by UltronAI** [comment]
@JiawangBian I visualize outputs of dispnet with different training settings. And I find that DispResNet can only predict a trivial result w/o gt for validation, while this bug won't occur in the case of training with gt or using VGG as backbone. It seems that DispResNet w/o gt is struggling in a local minimum. Have you ever meet this bug? And maybe it's similar to #2 that in my case (DispResNet w/o gt) the smoothness loss will soon drop to zero due to the trivial result even if its weight is 0.
![image](https://user-images.githubusercontent.com/14848863/68567195-5b7f6280-0493-11ea-8b65-742b53d5cf8c.png)

**Comment by JiawangBian** [comment]
1. This bug is regarless of using gt for validation, because gt is not used for training and not contributing any graident for avoiding terrible local minimum. 

2. It appears ramdonly. It may work well when you train that again without changing anything.

**Comment by UltronAI** [comment]
Yes, my statement is not accurate. But this phenomenon is not random in my tests. Every time I train my models on Cityscapes with ResNet as the backbone, I can only get naive results, so the results are terrible. However, it is much better when VGG is used as the backbone. And another confusing thing is that even when the outputs (depth and relative pose) are very bad, the warped image looks correct, which may be caused by jointly training. But what I didn't think clearly is whether this is an inherent problem with this joint training method, or is it just a bug in the code. Sorry to bother you many times. Thanks!
![image](https://user-images.githubusercontent.com/14848863/68917755-a2789b00-07a6-11ea-863c-93f296ed31f8.png)
![image](https://user-images.githubusercontent.com/14848863/68917791-c0de9680-07a6-11ea-8dde-7152d96a1d37.png)

**Comment by TopGun666** [comment]
> Yes, my statement is not accurate. But this phenomenon is not random in my tests. Every time I train my models on Cityscapes with ResNet as the backbone, I can only get naive results, so the results are terrible. However, it is much better when VGG is used as the backbone. And another confusing thing is that even when the outputs (depth and relative pose) are very bad, the warped image looks correct, which may be caused by jointly training. But what I didn't think clearly is whether this is an inherent problem with this joint training method, or is it just a bug in the code. Sorry to bother you many times. Thanks!
> ![image](https://user-images.githubusercontent.com/14848863/68917755-a2789b00-07a6-11ea-863c-93f296ed31f8.png)
> ![image](https://user-images.githubusercontent.com/14848863/68917791-c0de9680-07a6-11ea-8dde-7152d96a1d37.png)

Hi Feng @UltronAI ,

Is your depth visualization while training  written by yourself?  I haven't seen it in original code. Another question, have you test the posenet model trained by yourself? I test the model trained by myself, but it looks terrible.

regards

Yu

**Comment by UltronAI** [comment]
@TopGun666 Hi Yu. You can see codes for visualization in https://github.com/ClementPinard/SfmLearner-Pytorch. And the following picture shows the best results in my tests which are worse than those in the original paper.
![image](https://user-images.githubusercontent.com/14848863/69032319-2769eb80-0a17-11ea-992d-a3694da04eb8.png)

You can visualize your estimates and other information during training and see if it is also the case I mentioned.

**Comment by TopGun666** [comment]
Thanks for your reply.

**Comment by CongFang** [comment]
Where to download the pre-training model on CityScapes? Thanks!

**Comment by UltronAI** [comment]
@CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.

**Comment by CongFang** [comment]
> @CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.

I want the download link of the pre-training model on CityScapes. But, in README only have k_depth.tar and cs+k_depth.tar.

**Comment by JiawangBian** [comment]
> > @CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.
> 
> I want the download link of the pre-training model on CityScapes. But, in README only have k_depth.tar and cs+k_depth.tar.

Added.

**Comment by CongFang** [comment]
> > > @CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.
> > 
> > 
> > I want the download link of the pre-training model on CityScapes. But, in README only have k_depth.tar and cs+k_depth.tar.
> 
> Added.

Think you for reply. But, I think you uploaded the wrong model. Can you share the **depth** model pre-training model on CityScapes? As you mentioned in your paper: 'Also, we pre-train the network on CityScapes and fine-tune on KITTI, each for 200 epochs.'

-- | --

**Comment by CongFang** [comment]
> > > @CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.
> > 
> > 
> > I want the download link of the pre-training model on CityScapes. But, in README only have k_depth.tar and cs+k_depth.tar.
> 
> Added.

Think you for reply. But, I think you uploaded the wrong model. Can you share the **depth** model pre-training model on CityScapes? As you mentioned in your paper: 'Also, we pre-train the network on CityScapes and fine-tune on KITTI, each for 200 epochs.'

**Comment by JiawangBian** [comment]
> > > > @CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.
> > > 
> > > 
> > > I want the download link of the pre-training model on CityScapes. But, in README only have k_depth.tar and cs+k_depth.tar.
> > 
> > 
> > Added.
> 
> Think you for reply. But, I think you uploaded the wrong model. Can you share the **depth** model pre-training model on CityScapes? As you mentioned in your paper: 'Also, we pre-train the network on CityScapes and fine-tune on KITTI, each for 200 epochs.'

Updated

**Comment by CongFang** [comment]
> > > > > @CongFang you can find the download link in README. In my case, I want to pre-train models on Cityscapes by myself.
> > > > 
> > > > 
> > > > I want the download link of the pre-training model on CityScapes. But, in README only have k_depth.tar and cs+k_depth.tar.
> > > 
> > > 
> > > Added.
> > 
> > 
> > Think you for reply. But, I think you uploaded the wrong model. Can you share the **depth** model pre-training model on CityScapes? As you mentioned in your paper: 'Also, we pre-train the network on CityScapes and fine-tune on KITTI, each for 200 epochs.'
> 
> Updated

Thinks again!
